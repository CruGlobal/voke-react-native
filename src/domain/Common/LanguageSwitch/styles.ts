import theme from 'utils/theme';

const styles: { [key: string]: any } = {
  ...theme,
  settingOption: {
    color: '#fff',
    fontSize: theme.fontSizes.l,
  },

  langOption: {
    justifyContent: 'flex-start',
    alignSelf: 'center',
    paddingVertical: 10,
    paddingLeft: 10,
    paddingRight: 14,
    flexDirection: 'row',
    width: '100%',
  },
  langOptionText: {
    color: theme.colors.secondary,
    fontSize: theme.fontSizes.m,
  },
  langOptionCheckmark: {
    color: theme.colors.secondary,
    alignSelf: 'center',
    paddingLeft: theme.spacing.s,
    opacity: 0.5,
  },
};

export default styles;
