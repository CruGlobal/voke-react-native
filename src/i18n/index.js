import i18n from 'i18next';
import { reactI18nextModule } from 'react-i18next';
import * as moment from 'moment';
// Pull in all the moment locales
import 'moment/min/locales.min';

import { locale } from '../utils/common';

import translations from './locales/translations';

const languageDetector = {
  type: 'languageDetector',
  async: false,
  detect: () => locale,
  init: () => {},
  cacheUserLanguage: () => {},
};

// Set the date locale
moment.locale(locale);

export default i18n
  .use(languageDetector)
  .use(reactI18nextModule)
  .init({
    fallbackLng: 'en',

    // Use downloaded translations if available but use en-US from source to make development easier
    resources: translations,

    // have a common namespace used around the full app
    ns: ['common'],
    defaultNS: 'common',
    fallbackNS: 'common',

    interpolation: {
      escapeValue: false, // not needed for react as it does escape per default to prevent xss!
    },

    react: {
      wait: true,
      nsMode: 'fallback',
    },
  });
