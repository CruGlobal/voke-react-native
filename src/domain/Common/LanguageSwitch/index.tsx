import React, { ReactElement, useEffect, useState } from 'react';
import i18next from 'i18next';
import Select from 'domain/Common/Select';
import VokeIcon from 'components/VokeIcon';
import Text from 'components/Text';

import styles from './styles';

// Can't use import for json:
// https://github.com/microsoft/TypeScript/issues/24715
const languageCodes = require('i18n/langaugeCodes.json');
type Option = {
  label: string;
  selected?: boolean;
};

const LanguageSwitch = (): ReactElement => {
  const lang = languageCodes[i18next.language.substr(0, 2).toLowerCase()];
  const availableTranslations = i18next.languages;
  const [modalVisible, setModalVisible] = useState(false);
  const [selectOptions, setSelectOptions] = useState<Option[]>([]);

  useEffect(() => {
    // Prepare list of languages for selector removing 'en-US' used
    // for development purposes only.
    const appLang = i18next.language.toUpperCase();

    availableTranslations.map(lang => {
      const stepLang = lang.toUpperCase();
      if (stepLang !== 'EN-US') {
        if (
          appLang === stepLang ||
          (stepLang === 'EN' && appLang === 'EN-US')
        ) {
          setSelectOptions(current => [
            ...current,
            { label: lang, selected: true },
          ]);
        } else {
          setSelectOptions(current => [...current, { label: lang }]);
        }
      }
    });
  }, []);

  const newLangSelected = (option: Option): void => {
    // Change current language for app UI.
    const appLang = i18next.language.toUpperCase();
    const newSelectOptions: Option[] = selectOptions;
    selectOptions.map((value, index) => {
      if (
        value.label === option.label ||
        (appLang === 'EN-US' && option.label.toUpperCase() === 'EN')
      ) {
        // Mark selected language as current in the options object.
        newSelectOptions[index].selected = true;
        // Change current language in i18next instance:
        i18next.changeLanguage(value?.label);
      } else {
        delete selectOptions[index].selected;
      }
    });
    setSelectOptions(newSelectOptions);
  };

  return (
    <Select
      options={selectOptions}
      // Toggle button:
      toggleText={lang.name}
      toggleTestId="languageSwitchButton"
      toggleTextStyle={styles.settingOption}
      // Selector state change:
      isOpen={modalVisible}
      onOpen={() => setModalVisible(true)}
      onClose={() => setModalVisible(false)}
      onSelect={option => newLangSelected(option)}
      // Single option:
      optionEl={option => {
        return (
          <>
            {/* Item Label */}
            <Text style={styles.langOptionText}>
              {option.label.toUpperCase()}
            </Text>
            {/* Checkmark icon */}
            {option?.selected && (
              <VokeIcon
                name="checkmark-outline"
                size={10}
                style={styles.langOptionCheckmark}
                testID={'selected-' + option.label}
              />
            )}
          </>
        );
      }}
    />
  );
};

export default LanguageSwitch;
