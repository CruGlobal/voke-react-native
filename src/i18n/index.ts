import i18n, {
  LanguageDetectorModule,
  ResourceLanguage,
  Resource,
} from 'i18next';
import { NativeModules, Platform } from 'react-native';
import { initReactI18next, reactI18nextModule } from 'react-i18next';
import { findBestAvailableLanguage } from 'react-native-localize';
import oneSkyTranslations from './locales/translations.json';
import en_US from './locales/en-US.json';

const deviceLanguage =
          Platform.OS === 'ios'
            ? NativeModules.SettingsManager.settings.AppleLocale ||
              NativeModules.SettingsManager.settings.AppleLanguages[0] // iOS 13
            : NativeModules.I18nManager.localeIdentifier;
// import * as moment from 'moment';
// Pull in all the moment locales
// import 'moment/min/locales.min';

/* export const locale = DeviceInfo.getDeviceLocale()
  .substr(0, 2)
  .toLowerCase();
 */

// const languageDetector = {
//   type: 'languageDetector',
//   async: false,
//   detect: () => deviceLanguage.substr(0, 2).toLowerCase(),
//   init: () => {},
//   cacheUserLanguage: () => {},
// };

// Flatten OneSky translations to remove extra `translations` key
const mapOneSkyToResourceLanguages = (
  oneSkyTranslations: OneSkyTranslations,
) =>
  Object.entries(oneSkyTranslations).reduce(
    (resources, [localeTag, oneskyTranslation]) => ({
      ...resources,
      [localeTag]: oneskyTranslation.translation,
    }),
    {},
  );


const resourceLanguages = {
  // Use downloaded translations if available but use en-US from source to make development easier
  ...mapOneSkyToResourceLanguages(oneSkyTranslations),
  ...{ 'en-US': en_US as ResourceLanguage },
};

export const aliasLanguages = (
  aliases: { [alias: string]: string },
  translations: Resource,
) => ({
  ...translations,
  // Clone language translations to aliases if alias doesn't yet exist
  ...Object.entries(aliases).reduce(
    (aliasedTranslations, [alias, real]) => ({
      ...aliasedTranslations,
      // Keep existing translations if language exists or copy aliased language translations
      [alias]: translations[alias] || translations[real],
    }),
    {},
  ),
});

// Create alias mappings for base translations (key: alias languageTag, value: real languageTag)
export const createBaseLocaleAliases = (
  oneSkyTranslations: OneSkyTranslations,
) =>
  Object.keys(oneSkyTranslations).reduce(
    (baseAliases, languageTag) => ({
      ...baseAliases,
      [languageTag.split('-')[0]]: languageTag,
    }),
    {},
  );

const aliasedResourceLanguages: Resource = aliasLanguages(
  createBaseLocaleAliases(oneSkyTranslations),
  resourceLanguages,
);

const languageDetector: LanguageDetectorModule = {
  type: 'languageDetector',
  detect: () => {
    return (findBestAvailableLanguage(Object.keys(aliasedResourceLanguages)) || {})
      .languageTag;
  },
  init: () => {},
  cacheUserLanguage: () => {},
};

// Set the date locale
// moment.locale(locale);

export default i18n
  .use(languageDetector)
  .use(initReactI18next) // passes i18n down to react-i18next
  // .use(reactI18nextModule)
  .init({
    // lng: 'es',
    // whitelist: ['en', 'es', 'fr', 'pt' ],
    fallbackLng:  ['en', 'es', 'fr', 'pt' ],
    // debug: true,
    // Use downloaded translations if available but use en-US from source to make development easier
    resources: aliasedResourceLanguages,
    keySeparator: false, // we do not use keys in form messages.welcome
    // have a common namespace used around the full app
    ns: ['common'],
    defaultNS: 'common',
    fallbackNS: 'common',

    interpolation: {
      escapeValue: false, // not needed for react as it does escape per default to prevent xss!
    },

    /* react: {
      wait: true,
      nsMode: 'fallback',
    }, */
  });


  console.log( "üê∏ i18n.language:", i18n.language );