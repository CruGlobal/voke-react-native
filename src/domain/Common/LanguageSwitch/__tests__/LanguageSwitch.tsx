/**
 * @format
 */

import { Text, View } from 'react-native';
import React from 'react';
import { fireEvent, render, waitFor } from '@testing-library/react-native';

// import { toHaveStyle } from '@testing-library/jest-native'; - causes TS error.
// https://github.com/testing-library/jest-dom/issues/123
import '@testing-library/jest-native';
import i18n from 'i18next';

import LanguageSwitch from '../index';

// the modal component is automatically mocked by RN and apparently contains
// a bug which make the modal (and it's children) always visible in the test tree
// this is a hack which fix this issue
jest.mock('react-native/Libraries/Modal/Modal', () => {
  const Modal = jest.requireActual('react-native/Libraries/Modal/Modal');
  return (props) => <Modal {...props} />;
});

beforeEach(() => {
  // https://github.com/facebook/jest/issues/6434#issuecomment-525576660
  jest.useFakeTimers();
});

it('renders correctly', () => {
  const { queryByText, toJSON } = render(<LanguageSwitch />);

  expect(toJSON()).toMatchSnapshot();
  expect(queryByText(/English/i)).toBeTruthy();
  expect(queryByText(/FR/i)).not.toBeTruthy();
});

it('renders correct languages when open', async () => {
  const {
    getByTestId,
    queryByTestId,
    getByText,
    queryByText,
    toJSON,
    debug,
  } = render(<LanguageSwitch />);

  // FR language should be hidden first.
  expect(queryByText(/FR/i)).not.toBeTruthy();
  // Press on the language switch.
  fireEvent.press(getByTestId('languageSwitchButton'));
  // FR language present in the dropdown.
  expect(queryByText(/FR/i)).toBeTruthy();
  expect(queryByText(/ES/i)).toBeTruthy();
  expect(queryByText(/PT/i)).toBeTruthy();
  expect(queryByText(/\bEN\b/i)).toBeTruthy();
  expect(queryByTestId(/selected-fr/i)).not.toBeTruthy();
  expect(queryByTestId(/selected-en/i)).toBeTruthy();
  expect(toJSON()).toMatchSnapshot();
});

it('select new language when pressed', async () => {
  const {
    getByTestId,
    queryByTestId,
    getByText,
    queryByText,
    toJSON,
    debug,
  } = render(
    <View>
      <Text>{i18n.t('welcome:botMessageTitle')}</Text>
      <LanguageSwitch />
    </View>,
  );

  // FR language should be hidden first.
  expect(queryByText(/French/i)).not.toBeTruthy();
  // Check current string output.
  expect(queryByText(/Welcome to Voke!/i)).toBeTruthy();
  // Press on the language switch.
  fireEvent.press(getByTestId('languageSwitchButton'));
  // FR language present in the dropdown.
  fireEvent.press(getByText(/FR/i));
  // Dropdown should disappear.
  expect(queryByText(/ES/i)).not.toBeTruthy();
  // Current language label changed.
  expect(queryByText(/French/i)).toBeTruthy();
  // Check output change in i18n.
  expect(i18n.t('welcome:botMessageTitle') === 'Bienvenue Ã  Voke!');
  // Check checkmark inside of the modal.
  fireEvent.press(getByTestId('languageSwitchButton'));
  expect(queryByTestId(/selected-en/i)).not.toBeTruthy();
  expect(queryByTestId(/selected-fr/i)).toBeTruthy();
});
