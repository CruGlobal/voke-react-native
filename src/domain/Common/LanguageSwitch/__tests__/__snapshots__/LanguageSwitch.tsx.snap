// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correct languages when open 1`] = `
<View
  collapsable={false}
>
  <View
    accessible={true}
    focusable={true}
    forwardedRef={[Function]}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "opacity": 1,
      }
    }
    testID="languageSwitchButton"
  >
    <Text
      style={
        Array [
          Object {
            "color": "#1d1d26",
          },
          Object {
            "fontFamily": "TitilliumWeb-Regular",
          },
          Object {
            "color": "#fff",
            "fontSize": 17,
          },
        ]
      }
    >
      English
    </Text>
    <View
      style={
        Array [
          Object {
            "borderLeftColor": "transparent",
            "borderLeftWidth": 0,
            "borderRightColor": "white",
            "borderRightWidth": 1.5,
            "borderTopColor": "white",
            "borderTopWidth": 1.5,
            "height": 7,
            "transform": Array [
              Object {
                "translateX": 1.5,
              },
              Object {
                "translateY": -2.4,
              },
              Object {
                "rotate": "135deg",
              },
            ],
            "width": 7,
          },
          Object {
            "marginLeft": 10,
          },
        ]
      }
    />
  </View>
  <RCTModalHostView
    animationType="none"
    hardwareAccelerated={false}
    identifier={1}
    onRequestClose={[Function]}
    onStartShouldSetResponder={[Function]}
    presentationStyle="overFullScreen"
    style={
      Object {
        "position": "absolute",
      }
    }
    supportedOrientations={
      Array [
        "portrait",
        "portrait-upside-down",
        "landscape",
        "landscape-left",
        "landscape-right",
      ]
    }
    transparent={true}
  >
    <View
      collapsable={false}
      style={
        Array [
          Object {
            "flex": 1,
            "left": 0,
            "top": 0,
          },
          Object {
            "backgroundColor": "transparent",
          },
        ]
      }
    >
      <View
        pointerEvents="box-none"
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <View
          collapsable={true}
          pointerEvents="box-none"
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <View
            accessible={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
                Object {
                  "opacity": 0,
                },
              ]
            }
            testID="dropdownBackdrop"
          >
            <View
              onLayout={[Function]}
              style={
                Array [
                  Object {
                    "backgroundColor": "white",
                    "borderRadius": 4,
                    "position": "absolute",
                    "shadowColor": "black",
                    "shadowOffset": Object {
                      "height": 2,
                      "width": 0,
                    },
                    "shadowOpacity": 0.14,
                    "shadowRadius": 2,
                  },
                  Object {
                    "left": 0,
                    "top": 0,
                  },
                ]
              }
            >
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignSelf": "center",
                    "backgroundColor": "rgba(0,0,0,.07)",
                    "flexDirection": "row",
                    "justifyContent": "flex-start",
                    "opacity": 1,
                    "paddingLeft": 10,
                    "paddingRight": 14,
                    "paddingVertical": 10,
                    "width": "100%",
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#1d1d26",
                      },
                      Object {
                        "fontFamily": "TitilliumWeb-Regular",
                      },
                      Object {
                        "color": "#186078",
                        "fontSize": 15,
                      },
                    ]
                  }
                >
                  EN
                </Text>
                <Text
                  allowFontScaling={false}
                  style={
                    Array [
                      Object {
                        "color": undefined,
                        "fontSize": 10,
                      },
                      Object {
                        "alignSelf": "center",
                        "color": "#186078",
                        "opacity": 0.5,
                        "paddingLeft": 8,
                      },
                      Object {
                        "fontFamily": "icoMoon",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                  testID="selected-en"
                >
                  î¤´
                </Text>
              </View>
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignSelf": "center",
                    "backgroundColor": "transparent",
                    "flexDirection": "row",
                    "justifyContent": "flex-start",
                    "opacity": 1,
                    "paddingLeft": 10,
                    "paddingRight": 14,
                    "paddingVertical": 10,
                    "width": "100%",
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#1d1d26",
                      },
                      Object {
                        "fontFamily": "TitilliumWeb-Regular",
                      },
                      Object {
                        "color": "#186078",
                        "fontSize": 15,
                      },
                    ]
                  }
                >
                  ES
                </Text>
              </View>
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignSelf": "center",
                    "backgroundColor": "transparent",
                    "flexDirection": "row",
                    "justifyContent": "flex-start",
                    "opacity": 1,
                    "paddingLeft": 10,
                    "paddingRight": 14,
                    "paddingVertical": 10,
                    "width": "100%",
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#1d1d26",
                      },
                      Object {
                        "fontFamily": "TitilliumWeb-Regular",
                      },
                      Object {
                        "color": "#186078",
                        "fontSize": 15,
                      },
                    ]
                  }
                >
                  FR
                </Text>
              </View>
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignSelf": "center",
                    "backgroundColor": "transparent",
                    "flexDirection": "row",
                    "justifyContent": "flex-start",
                    "opacity": 1,
                    "paddingLeft": 10,
                    "paddingRight": 14,
                    "paddingVertical": 10,
                    "width": "100%",
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#1d1d26",
                      },
                      Object {
                        "fontFamily": "TitilliumWeb-Regular",
                      },
                      Object {
                        "color": "#186078",
                        "fontSize": 15,
                      },
                    ]
                  }
                >
                  PT
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTModalHostView>
</View>
`;

exports[`renders correctly 1`] = `
<View
  collapsable={false}
>
  <View
    accessible={true}
    focusable={true}
    forwardedRef={[Function]}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "opacity": 1,
      }
    }
    testID="languageSwitchButton"
  >
    <Text
      style={
        Array [
          Object {
            "color": "#1d1d26",
          },
          Object {
            "fontFamily": "TitilliumWeb-Regular",
          },
          Object {
            "color": "#fff",
            "fontSize": 17,
          },
        ]
      }
    >
      English
    </Text>
    <View
      style={
        Array [
          Object {
            "borderLeftColor": "transparent",
            "borderLeftWidth": 0,
            "borderRightColor": "white",
            "borderRightWidth": 1.5,
            "borderTopColor": "white",
            "borderTopWidth": 1.5,
            "height": 7,
            "transform": Array [
              Object {
                "translateX": 1.5,
              },
              Object {
                "translateY": -2.4,
              },
              Object {
                "rotate": "135deg",
              },
            ],
            "width": 7,
          },
          Object {
            "marginLeft": 10,
          },
        ]
      }
    />
  </View>
</View>
`;
